### PR TITLE
ISPN-13052 Hot Rod Java client doc updates

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-13052

### CODE REVIEW

**Review by pruivo — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_java_clients.adoc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+----
+
+[discrete]
+== Verifying connections
+
+You can check that Hot Rod client connections are active by examining the transport with the `getChannelFactory()` method, as in the following example:
+
+----
+RemoteCacheManager rcm = new RemoteCacheManager(configuration);
+      boolean started = rcm.getChannelFactory().getNumActive() > 0;
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/proc_configuring_hotrod_java_clients.adoc:45]: This actually came from the QE team. Maybe let's just remove it if it doesn't make that much sense.

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_connection_pools.adoc [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
+.ConfigurationBuilder
+[source,java,options="nowrap",subs=attributes+,role="primary"]
 ----
 include::code_examples/HotRodConnectionPools.java[]
 ----
+
+.hotrod-client.properties
+[source,options="nowrap",subs=attributes+,role="secondary"]
+----
+include::config_examples/hotrod-client.properties[]
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/proc_configuring_hotrod_connection_pools.adoc:21]: nice catch, thanks.

File: documentation/src/main/asciidoc/topics/proc_configuring_hotrod_connection_pools.adoc [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
 = Configuring Connection Pools
-Hot Rod Java clients keep pools of persistent connections to {brandname}
-servers to reuse TCP connections instead of creating them on each request.
+Hot Rod Java clients keep pools of persistent connections to {brandname} servers to reuse TCP connections instead of creating them on each request.
 
-Clients use asynchronous threads that check the validity of connections by
-iterating over the connection pool and sending pings to {brandname} servers.
-This improves performance by finding broken connections while they are idle in
-the pool, rather than on application requests.
+Clients use asynchronous threads that check the validity of connections by iterating over the connection pool and sending pings to {brandname} servers.
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/proc_configuring_hotrod_connection_pools.adoc:5]: removed this paragraph, cheers.

File: documentation/src/main/asciidoc/topics/code_examples/HotRodForceReturn.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+cache.put("aKey", "initialValue");
+assert null == cache.put("aKey", "aValue");
+//Use the `FORCE_RETURN_VALUE` flag to return previous values for keys.
+assert "aValue".equals(cache.withFlags(Flag.FORCE_RETURN_VALUE).put("aKey",
+   "newValue"));
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/code_examples/HotRodForceReturn.java:5]: less is more

File: documentation/src/main/asciidoc/topics/code_examples/HotRodCreatePerCache.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+ConfigurationBuilder builder = new ConfigurationBuilder();
+builder.remoteCache("another-cache")
+       .configuration("<distributed-cache name=\"another-cache\"/>");
+builder.remoteCache("my.other.cache")
+       .configurationURI(URI.create("file:/path/to/infinispan.xml"));
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/code_examples/HotRodCreatePerCache.java]: @pruivo I think showing `File` is better. I'm going to add that. thanks.

File: documentation/src/main/asciidoc/topics/proc_creating_caches_hotrod.adoc [hunk 6]
```diff
@@ -1,78 +1,48 @@
...
+----
+
+[IMPORTANT]
+====
+When using `hotrod-client.properties` with cache names that contain the `.` character, you must enclose the cache name in square brackets as in the preceding example.
+====
+
+You can also create remote caches through through the `RemoteCacheManager` API in other ways, such as the following example that adds a cache configuration with the `XMLStringConfiguration()` method and then calls the `getOrCreateCache()` method.
+
+However, {brandname} does not recommend this approach because you need to include XML directly in your Java code.
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/proc_creating_caches_hotrod.adoc:30]: I knew I was going to get comments on this one :laughing: I'll reword this

File: documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc [hunk 7]
```diff
@@ -1,23 +1,25 @@
-[id='request_balancing']
+[id='request-balancing_{context}']
 = Request Balancing
-Clients that use topology-aware intelligence use request balancing for all
-requests. The default balancing strategy is round-robin, so topology-aware
-clients always send requests to servers in round-robin order.
+Clients that use topology-aware intelligence use request balancing for all requests. The default balancing strategy is round-robin, so topology-aware clients always send requests to {brandname} Server in round-robin order.
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc:3]: Can we just reduce it down to "Hot Rod Java clients use request balancing for all requests"?

[the remark this answers is not in the document]
Comment (pruivo) [reply documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc]: probably would say:
`Clients that use basic or topology-aware intelligence use request balancing for all requests. Clients using hash-distribution-aware intelligence send the request to the node that stores the key and fallbacks to request balancing if this node fails to reply. The default balancing strategy...`

BTW, the section before (1.1.1 Client Intelligence) is missing information on how to configure the client intelligence :) and, if possible, add the word `Intelligence to the intelligence names.

File: documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc [hunk 8]
```diff
@@ -46,6 +46,12 @@ However, when using the asynchronous backup strategy, {brandname} can automatica
 When it detects that a backup location has come back online, and the network connection is stable, {brandname} initiates bi-directional state transfer between backup locations.
 For example, {brandname} simultaneously transfers state from **LON** to **NYC** and **NYC** to **LON**.
 
+[NOTE]
+====
+To avoid temporary network disconnects triggering state transfer operations, there are two conditions that backup locations must meet to go offline.
+The status of a backup location must be offline and it must not be included in the cross-site view with JGroups RELAY2.
+====
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc:53]: @pruivo Hey I hope it's ok to be a little bit bold and sneak this in here. We discussed the two conditions (jgroups + infinispan) that need to be met before state transfer is triggered when sites come back online. I changed the wording a little. Hopefully didn't alter meaning. :grin:

File: documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc [hunk 9]
```diff
@@ -1,23 +1,28 @@
...
 
 .Custom balancing policies
-You can implement link:{javadocroot}/org/infinispan/client/hotrod/FailoverRequestBalancingStrategy.html[FailoverRequestBalancingStrategy] and specify your class in your
-`hotrod-client.properties` configuration with the following property:
 
-`infinispan.client.hotrod.request_balancing_strategy`
+You can use custom `FailoverRequestBalancingStrategy` implementations if you add your class in the Hot Rod client configuration in one of the following ways:
+
+* `balancingStrategy()` in your `ConfigurationBuilder`.
+* `infinispan.client.hotrod.request_balancing_strategy` in `hotrod-client.properties`.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc:24]: put some example instead? wdyt?

```
ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
clientBuilder.addServer()
              .host("127.0.0.1")
              .port(11222)
              .balancingStrategy(new MyCustomBalancingStrategy())
```
`infinispan.client.hotrod.request_balancing_strategy=my.package.MyCustomBalancingStrategy`

File: documentation/src/main/asciidoc/topics/code_examples/HotRodClientInstance.java [hunk 10]
```diff
@@ -1,10 +1,10 @@
...
-      .enable()
-      .jmxDomain("org.example")
-  .addServer()
-      .host("127.0.0.1")
-      .port(11222);
-RemoteCacheManager rmc = new RemoteCacheManager(cb.build());
+ConfigurationBuilder builder = new ConfigurationBuilder();
+builder.addServer()
+         .host("127.0.0.1")
+         .port(ConfigurationProperties.DEFAULT_HOTROD_PORT)
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/code_examples/HotRodClientInstance.java:4]: it is missing `192.0.2.0:11222`
the URI example has 2 servers and the `hotrod-client.properties` also has 2 servers.

**Review by pruivo — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc [hunk 11]
```diff
@@ -1,23 +1,28 @@
...
 
 [source,java]
 ----
 include::code_examples/RequestBalancing.java[]
 ----
 
-Clients that use distribution-aware intelligence use request balancing only
-for failed requests. When requests fail, distribution-aware clients retry the
-request on the next available server.
+Clients that use distribution-aware intelligence use request balancing only for failed requests.
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_hotrod_request_balancing.adoc:16]: bah. fixed.

File: documentation/src/main/asciidoc/topics/code_examples/RequestBalancing.java [hunk 12]
```diff
@@ -1,12 +1,13 @@
-CacheContainer cacheContainer = new RemoteCacheManager();
-Cache<String, String> cache = cacheContainer.getCache();
+// Connect to the {brandname} cluster
```
[the remark this answers is not in the document]
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/code_examples/RequestBalancing.java:1]: Need to add `subs=attributes+` to the code block, which I thought I did when I added the variable but I guess not. thanks for catching this.

**Review by pruivo — APPROVED** [review]

**Comment by ryanemerson** [comment]
Thanks @oraNod

**Pull request merged**